### ISSUE TITLE
Example of misformatting with the current clang-format

### ISSUE THREAD

**Issue by sbastrakov** [ComputationalRadiationPhysics/contributing#48]
[This piece in PIConGPU](https://github.com/ComputationalRadiationPhysics/picongpu/blob/f6f15f7e4ff96f2ec3fc7630b7ba021b8ff83a9b/include/picongpu/fields/MaxwellSolver/YeePML/YeePML.kernel#L398-L400) seems to have been misunderstood and therefore misformatted by `clang-format`. I am not sure exactly why.

Admittedly, it is rather and probably too complicated anyhow. What is going on there:

- `ForEachIdx<IdxConfig<numCellsPerSuperCell, numWorkers>>` is a functor type
- We instantiate an object of this type by passing `workerIdx` to its constructor, so the curly brace at the end of line 398 is call for constructor, not start of a block
- Then we immediately call `operator()` of the freshly constructed object in the same expression
- And pass a lambda defined just there as its argument

cc @j-stephan

**Comment by bernhardmgruber** [comment]
The formatting looks correct to me. What is incorrectly formatted?

**Comment by sbastrakov** [comment]
Well, it formatted as
```
Type{
     constructor parameter}(...;
```

, I do not think it is correct.

**Comment by psychocoderHPC** [comment]
What is the result if we use `()` for the constructor?

**Comment by j-stephan** [comment]
I'm not sure what the preferable way of formatting should look like.

The current way isn't pretty:

```c++
ForEachIdx<IdxConfig<numCellsPerSuperCell, numWorkers>>{
    workerIdx}([&](uint32_t const linearIdx, uint32_t const) {
```

But this is even worse IMO:

```c++
ForEachIdx<IdxConfig<numCellsPerSuperCell, numWorkers>>{workerIdx}
    ([&](uint32_t const linearIdx, uint32_t const) {
```

We could also break somewhere in the lambda parameter list but that is awful, too. I guess this is just one of those cases where there is no good solution.

**Comment by sbastrakov** [comment]
I agree the code is just very complicated and that's the main problem. But how it got formatted is in my opinion not just visually iffy, but more importantly actively misleading, as from the first look it seems to have a new code starting, which is not the case,

**Comment by j-stephan** [comment]
Would it be possible to have an `ForEachIdx` object? This way it could be broken into two lines:

```c++
auto loop = ForEachIdx<IdxConfig<numCellsPerSuperCell, numWorkers>>{workerIdx};
loop([&](uint32_t const linearIdx, uint32_t const)
{
    /* body goes here */
});
```

**Comment by sbastrakov** [comment]
Yes, I think that would have been better to do in the first place.
